File: source/javascript/coin_change.js
// Coin Change Problem (Minimum Coins)
function coinChange(coins, amount) {
    const dp = new Array(amount + 1).fill(amount + 1);
    dp[0] = 0;
    for (let i = 1; i <= amount; i++) {
        for (const coin of coins) {
            if (coin <= i) {
                const candidate = dp[i - coin] + 1;
                if (candidate < dp[i]) dp[i] = candidate;
            }
        }
    }
    return dp[amount] > amount ? -1 : dp[amount];
}

const coins = [1, 5, 10, 25];
const amount = 36;
console.log(`min coins: ${coinChange(coins, amount)}`);

File: source/rust/coin_change.rs
fn coin_change(coins: &[i32], amount: i32) -> i32 {
    let amount = amount as usize;
    let mut dp = vec![amount as i32 + 1; amount + 1];
    dp[0] = 0;
    for i in 1..=amount {
        for &coin in coins {
            if coin as usize <= i {
                let candidate = dp[i - coin as usize] + 1;
                if candidate < dp[i] { dp[i] = candidate; }
            }
        }
    }
    if dp[amount] > amount as i32 { -1 } else { dp[amount] }
}

fn main() {
    let coins = [1, 5, 10, 25];
    println!("min coins: {}", coin_change(&coins, 36));
}
